File: etc/problem69.lua

local function gcd(a, b)
  while b ~= 0 do
    a, b = b, a % b
  end
  return a
end

local upperlimit = 1000000

local primes = {}
for n=1,upperlimit do
  primes[n] = true
end
primes[1] = false

for i=2,upperlimit do
  if primes[i] then
    local n = i + i
    while n < upperlimit do
      primes[n] = false
      n = n + i
    end
  end
end

local function phi(n)
  if primes[n] then
    return n - 1
  end
  local prod = n
  for p=2,n do
    if primes[p] then
      if n % p == 0 then
        prod = prod * (1 - 1 / p)
      end
    end
  end
  return prod
end

-- local function phi(n)
--   if primes[n] then
--     return n - 1
--   end
--   local prod = n
--   for p=2,math.sqrt(n) do
--     if primes[p] then
--       if n % p == 0 then
--         prod = prod * (1 - 1 / p)
--         if p * p ~= n and primes[n / p] then
--           prod = prod * (1 - p / n)
--         end
--       end
--     end
--   end
--   return prod
-- end

local max = 0
local maxratio = 0

for n=1,1000000 do

  local val = phi(n)
  local ratio = n / val
  if ratio > maxratio then
    max = n
    maxratio = ratio
  end

end

print(max)

File: problem69.m

#import <Foundation/Foundation.h>
#import <math.h>
#import <stdio.h>

double phi[1000001];

int main(int argc, const char* argv[]) {
  NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
  int n, i;
  for (i = 0; i <= 1000000; i++) {
    phi[i] = i;
  }
  for (i = 2; i <= 1000000; i++) {
    if (phi[i] == i) {
      n = i;
      while (n <= 1000000) {
        phi[n] = phi[n] * (1 - 1 / (double)i);
        n += i;
      }
    }
  }
  int max = 0;
  double maxratio = 0;
  for (n = 1; n <= 1000000; n++) {
    int val = phi[n];
    double ratio = ((double)n) / val;
    if (ratio > maxratio) {
      max = n;
      maxratio = ratio;
    }
  }
  NSString* result = [NSString stringWithFormat:@"%d", max];
  puts(result.UTF8String);
  [pool release];
  return 0;
}
